python/ray/tune/trial_runner.py

```diff
@@ -1310,47 +1280,31 @@ def _process_stop_requests(self):
-
-        Trials may be stopped at any time. If trial is in state PENDING
-        or PAUSED, calls `on_trial_remove`  for scheduler and
-        `on_trial_complete() for search_alg.
-        Otherwise waits for result for the trial and calls
-        `on_trial_complete` for scheduler and search_alg if RUNNING.
+        """The canonical implementation of stopping a trial.
+
+        Trials may be in any external status when this function is called.
+        If trial is in state PENDING or PAUSED, calls `on_trial_remove` for
+        scheduler and `on_trial_complete()` for search_alg.
+        If trial is in state RUNNING, calls `on_trial_complete` for scheduler
+        and search_alg if RUNNING. Caller to ensure that there is no
+        outstanding future to be handled for the trial. If there is, the future
+        would be discarded.
         """
-        error = False
-        error_msg = None
-
         if trial.status in [Trial.ERROR, Trial.TERMINATED]:
             return
         elif trial.status in [Trial.PENDING, Trial.PAUSED]:
             self._scheduler_alg.on_trial_remove(self, trial)
             self._search_alg.on_trial_complete(trial.trial_id)
-            self._callbacks.on_trial_complete(
-                iteration=self._iteration, trials=self._trials, trial=trial)
         elif trial.status is Trial.RUNNING:
-            try:
-                results = self.trial_executor.fetch_result(trial)
-                result = results[-1]
-                trial.update_last_result(result)
-                self._scheduler_alg.on_trial_complete(self, trial, result)
-                self._search_alg.on_trial_complete(
-                    trial.trial_id, result=result)
-                self._callbacks.on_trial_complete(
-                    iteration=self._iteration,
-                    trials=self._trials,
-                    trial=trial)
-            except Exception:
-                error_msg = traceback.format_exc()
-                logger.exception("Error processing event.")
-                self._scheduler_alg.on_trial_error(self, trial)
-                self._search_alg.on_trial_complete(trial.trial_id, error=True)
-                self._callbacks.on_trial_error(
-                    iteration=self._iteration,
-                    trials=self._trials,
-                    trial=trial)
-                error = True
-        self.trial_executor.stop_trial(trial, error=error, error_msg=error_msg)
+            # By this time trial.last_result should have been updated already.
+            self._scheduler_alg.on_trial_complete(
+                self, trial, flatten_dict(trial.last_result))
+            self._search_alg.on_trial_complete(
+                trial.trial_id, result=flatten_dict(trial.last_result))
+        self._callbacks.on_trial_complete(
+            iteration=self._iteration, trials=self._trials, trial=trial)
+        self.trial_executor.export_trial_if_needed(trial)
+        self.trial_executor.stop_trial(trial)
```

Comment:
Should we still have this in a try-catch block? Otherwise the whole tune run will fail even though we could just have run into a logging issue. E.g. when an API comes down during a long run, we wouldn't want Tune to crash.

Reply:
I see. That's a good point. What would be the expected behavior? We can either do nothing, or mark the trial as ERROR.

Reply:
Current behavior is to set to ERROR - let's keep that for now!